examples/tpm2-seal-unseal/main.go

```diff
@@ -0,0 +1,281 @@
+	}
+	fmt.Printf("Sealed data: 0x%x\n", privateArea)
+
+	// Load the sealed data into the TPM.
+	objectHandle, _, err := tpm2.Load(rwc, srkHandle, srkPassword, publicArea, privateArea)
+	if err != nil {
+		return append(errs, fmt.Errorf("unable to load data: %v", err))
+	}
+	defer func() {
+		if err := tpm2.FlushContext(rwc, objectHandle); err != nil {
+			errs = append(errs, fmt.Errorf("unable to flush object handle %q: %v", objectHandle, err))
+		}
+	}()
+	fmt.Printf("Loaded sealed data with handle: 0x%x\n", objectHandle)
+
+	// Unseal the data
+	unsealedData, unsealErr, otherErrs := unseal(rwc, pcr, objectPassword, objectHandle)
+	if unsealErr != nil {
+		errs = append(errs, fmt.Errorf("unable to unseal data: %v", unsealErr))
+	}
+	for otherErr := range otherErrs {
+		errs = append(errs, fmt.Errorf("error while unsealing data: %v", otherErr))
+	}
+	if len(errs) > 0 {
+		return
+	}
+	fmt.Printf("Unsealed data: 0x%x\n", unsealedData)
+
+	// Try to unseal the data with the wrong password
+	_, unsealErr, otherErrs = unseal(rwc, pcr, "wrong-password", objectHandle)
+	if unsealErr == nil && len(otherErrs) == 0 {
+		errs = append(errs, fmt.Errorf("expected unsealing with wrong password to fail but it succeeded"))
+	}
+	for otherErr := range otherErrs {
+		errs = append(errs, fmt.Errorf("error while unsealing data: %v", otherErr))
+	}
+	if len(errs) > 0 {
+		return
+	}
+	// Note that this can be easily fooled since we don't verify that the error was
+	// due specifically to policy.
+	fmt.Printf("As expected, unsealing with the wrong password failed with an error. Error was: %v\n", unsealErr)
+
+	// Extend the PCR
+	if err = tpm2.PCREvent(rwc, tpmutil.Handle(pcr), []byte{1}); err != nil {
+		return append(errs, fmt.Errorf("unable to extend PCR: %v", err))
+	}
+	fmt.Printf("Extended PCR %d\n", pcr)
+
+	// Note the new value of the pcr
+	pcrVal, err = readPCR(rwc, pcr)
+	if err != nil {
+		return append(errs, fmt.Errorf("unable to read PCR: %v", err))
+	}
+	fmt.Printf("PCR %d value: 0x%x\n", pcr, pcrVal)
+
+	// Try to unseal the data with the PCR in the wrong state
+	_, unsealErr, otherErrs = unseal(rwc, pcr, objectPassword, objectHandle)
+	if unsealErr == nil && len(otherErrs) == 0 {
+		errs = append(errs, fmt.Errorf("expected unsealing with wrong PCR state to fail but it succeeded"))
+	}
+	for otherErr := range otherErrs {
+		errs = append(errs, fmt.Errorf("error while unsealing data: %v", otherErr))
+	}
+	if len(errs) > 0 {
+		return
+	}
+	// Note that this can be easily fooled since we don't verify that the error was
+	// due specifically to policy.
+	fmt.Printf("As expected, unsealing after extending the PCR failed with an error. Error was: %v\n", unsealErr)
+
+	return
+}
+
+// Returns the unsealed data, any error due to the Unseal command, and any other errors
+func unseal(rwc io.ReadWriteCloser, pcr int, objectPassword string, objectHandle tpmutil.Handle) (data []byte, unsealErr error, otherErrs []error) {
```

Comment:
Return just `([]byte, error)`. The separation of errors is too complicated

Reply:
The separation of errors allows the caller to determine whether the call behaved as expected or not (see usage of the returned errors above).

Reply:
You only separate these because your example became an integration test.
If you remove the code that tests failure scenarios, the caller doesn't care what kind of error happen.